### ISSUE TITLE
set_element_information throws exception: object_depth returns -1

### ISSUE THREAD

**Issue by jsa34** [gaozhao1989/pyjab#4]
See below error:
```
>       jab.enter_textbox_details_by_label("Name Required", input_name)
ar_test.py:36: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\utils\jab_actions.py:21: in enter_textbox_details_by_label
    element.send_text(input_text)
..\..\pyjab\jabelement.py:864: in send_text
    self.clear()
..\..\pyjab\jabelement.py:301: in clear
    self.click()
..\..\pyjab\jabelement.py:284: in click
    self.set_element_information()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <pyjab.jabelement.JABElement object at 0x000001FD110C8BB0>
    def set_element_information(self) -> None:
        """Set JABElement information(attributes) from internal jab funcs
    
        Raises:
            JABException: None value of vmid, hwnd or accessible_context
            JABException: Get Accessible Context Info error(internal jab func error)
            JABException: Get Object Depth error(internal jab func error)
        """
        if not any([self.vmid, self.hwnd, self.accessible_context]):
            raise JABException(
                "JABElement attributes should have vmid, hwnd and accessible_context"
            )
        info = AccessibleContextInfo()
        result = self.bridge.getAccessibleContextInfo(
            self.vmid, self.accessible_context, byref(info)
        )
        if result == 0:
            raise JABException(self.int_func_err_msg.format("GetAccessibleContextInfo"))
        object_depth = self.bridge.getObjectDepth(self.vmid, self.accessible_context)
        if object_depth == -1:
>           raise JABException(self.int_func_err_msg.format("getObjectDepth"))
E           pyjab.common.exceptions.JABException: ("Java Access Bridge func 'getObjectDepth' error", None)
..\..\pyjab\jabelement.py:954: JABException

```

**Comment by gaozhao1989** [comment]
Hi @jsa34,

Could you please provide the element information from func "get_element_information" after you specific Java Element.
General speaking, if object depth equal -1, that's means not specific the correct Java Accessible Context.

**Comment by jsa34** [comment]
Sure - apologies, I meant to include the element info - This is what Pycharm evaluates the element as before send_text is called:

```
result = {JABElement} <pyjab.jabelement.JABElement object at 0x000001EF7D96C9A0>
 accessible_action = {bool} False
 accessible_component = {bool} True
 accessible_context = {JOBJECT64} <JOBJECT64 object at 0x000001EF0053CAC0>
 accessible_interfaces = {bool} False
 accessible_selection = {bool} False
 accessible_text = {bool} True
 bounds = {dict: 4} {'x': 155, 'y': 168, 'height': 24, 'width': 480}
 bridge = {CDLL} <CDLL 'C:\Program Files\Java\jdk1.8.0_192\jre\bin\WindowsAccessBridge-64.dll', handle 7ff8d26b0000 at 0x1ef7d6f9f70>
 children_count = {int} 0
 description = {str} 'Name Required'
 hwnd = {c_void_p} c_void_p(526320)
 index_in_parent = {int} 2
 int_func_err_msg = {str} 'Java Access Bridge func \'{}\' error'
 location = {str} 'Traceback (most recent call last):\n  File "C:\\Users\\Jason\\AppData\\Local\\JetBrains\\Toolbox\\apps\\PyCharm-C\\ch-0\\211.7628.24\\plugins\\python-ce\\helpers\\pydev\\_pydevd_bundle\\pydevd_resolver.py", line 178, in _getPyDictionary\n    attr = getattr(
 name = {str} 'Name Required'
 object_depth = {int} 18
 parent = {str} 'Traceback (most recent call last):\n  File "C:\\Users\\Jason\\AppData\\Local\\JetBrains\\Toolbox\\apps\\PyCharm-C\\ch-0\\211.7628.24\\plugins\\python-ce\\helpers\\pydev\\_pydevd_bundle\\pydevd_resolver.py", line 178, in _getPyDictionary\n    attr = getattr(
 role = {str} 'text'
 role_en_us = {str} 'text'
 shortcut_keys = {ShortcutKeys} <pyjab.common.shortcutkeys.ShortcutKeys object at 0x000001EF7D79DE20>
 size = {str} 'Traceback (most recent call last):\n  File "C:\\Users\\Jason\\AppData\\Local\\JetBrains\\Toolbox\\apps\\PyCharm-C\\ch-0\\211.7628.24\\plugins\\python-ce\\helpers\\pydev\\_pydevd_bundle\\pydevd_resolver.py", line 178, in _getPyDictionary\n    attr = getattr(
 states = {list: 6} ['enabled', 'focusable', 'visible', 'showing', 'editable', 'single line']
 states_en_us = {list: 6} ['enabled', 'focusable', 'visible', 'showing', 'editable', 'single line']
 text = {str} ''
 vmid = {c_long} c_long(132672)
 win32_utils = {Win32Utils} <pyjab.common.win32utils.Win32Utils object at 0x000001EF7D7120D0>
 xpath_parser = {XpathParser} <pyjab.common.xpathparser.XpathParser object at 0x000001EF7D79DAF0>
```

...and get_element_information() shows:
`{'name': 'Name Required', 'description': 'Name Required', 'role': 'text', 'role_en_us': 'text', 'states': ['enabled', 'focusable', 'visible', 'showing', 'editable', 'single line'], 'states_en_us': ['enabled', 'focusable', 'visible', 'showing', 'editable', 'single line'], 'bounds': {'x': 155, 'y': 168, 'height': 24, 'width': 480}, 'object_depth': 18, 'index_in_parent': 2, 'children_count': 0, 'accessible_component': True, 'accessible_action': False, 'accessible_selection': False, 'accessible_text': True, 'text': ''}`

**Comment by jsa34** [comment]
It's odd, because BEFORE clicking, it has the info, including object_depth.

However, I notice .click func resets the element:

        self.win32_utils._set_window_foreground(hwnd=self.hwnd.value)
        self.set_element_information()

And when doing this, this is where it returns the error above

**Comment by jsa34** [comment]
It seems to be resolved by removing the line:
self.set_element_information()

**Comment by gaozhao1989** [comment]
Hi @jsa34,

Actually func "set_element_informaion" before action "click" is refresh the JABElement information, it because pyjab need to known the element "bounds" to avoid element change after some "actions".
This error looks like occurred  by after recall the func "set_element_informaion", but this is internal JAB func error.
I have retest in my side, unfortunately this does not reproduced.

Need more information about the error in your scenarios, would you please help  with below information? 
1. Is JAB internal func "getObjectDepth" will always return -1 after multiple usage
2. Is pyjab func "set_element_informaion" will failed after multiple usage

Other solutions for you(need you to handle some JAB internal function)
1. Check with your local environment of JDK and runtime, try with x86 version(not fully test x64 version before, not sure impact by this)
2. Try JAB internal func "setTextContents" for text input and "doAccessibleActions" in your side if the BUG for JAB is not impact for your side

**Comment by jsa34** [comment]
Q: Is JAB internal func "getObjectDepth" will always return -1 after multiple usage:
Q: Is pyjab func "set_element_information" will failed after multiple usage
A: It appears to be occurring consistently when calling "click()" function where a known element is re-calling set_element_information().
I stuck a breakpoint in self.set_element_information() when it's setting the initial element list, and rerunning self.set_element_information() doesn't cause an error.
Running self.bridge.getObjectDepth(self.vmid, self.accessible_context) multiple times at breakpoint returns the same value consistently (no -1)

I will try the second two suggestions

**Comment by jsa34** [comment]
For completeness, I realised I hadn't included the actual code used to generate the element locator prior to sending .send_keys():

```
        _index = index - 1 if index else 0
        all_elements = self.driver.find_elements(by=By.XPATH, value="//text")
        element: JABElement = [
            e
            for e in all_elements
            if (e.name.startswith(label))
            and (e.role == "text")
            and ("visible" in e.states)
        ][_index]
        element.send_text(input_text) << This is where the error occurs
        time.sleep(1)

```

**Comment by jsa34** [comment]
After realising I was using an older version of Java 8 (still 64-bit), I reran and now get the following error when trying to click:

```
self = <pyjab.jabelement.JABElement object at 0x0000020D283D66D0>

    def set_element_information(self) -> None:
        """Set JABElement information(attributes) from internal jab funcs
    
        Raises:
            JABException: None value of vmid, hwnd or accessible_context
            JABException: Get Accessible Context Info error(internal jab func error)
            JABException: Get Object Depth error(internal jab func error)
        """
        if not any([self.vmid, self.hwnd, self.accessible_context]):
            raise JABException(
                "JABElement attributes should have vmid, hwnd and accessible_context"
            )
        info = AccessibleContextInfo()
        result = self.bridge.getAccessibleContextInfo(
            self.vmid, self.accessible_context, byref(info)
        )
        if result == 0:
>           raise JABException(self.int_func_err_msg.format("GetAccessibleContextInfo"))
E           pyjab.common.exceptions.JABException: ("Java Access Bridge func 'GetAccessibleContextInfo' error", None)

..\..\pyjab\jabelement.py:949: JABException

```

I am now trying the 32-bit suggestion

**Comment by jsa34** [comment]
Interesting, running a second time with no changes resulted in the same error as before, and seemed to return happily at result == 0... Curious

**Comment by jsa34** [comment]
Just tried with 32-bit python and java, and still same response...

**Comment by jsa34** [comment]
So... What consistently seems to be the difference to make it work is:

```
        # _index = index - 1 if index else 0
        # all_elements = self.driver.find_elements(by=By.XPATH, value="//text")
        # element: JABElement = [
        #     e
        #     for e in all_elements
        #     if (e.name.startswith(label))
        #     and (e.role == "text")
        #     and ("visible" in e.states)
        # ][_index]
        # element.send_text(input_text)
        # time.sleep(1)
        self.driver.find_element(by=By.XPATH, value=f"//[@name='{label}']").send_text(input_text)

```

Commented out old code and the one that works.

This works when I have a uniqueness, but there are a number of items that don't so I need to either list them and retrieve by index or do this with the xpath (the parser doesn't currently allow xpaths like this, though: (//push button[@name='Name'])[2] )

So it looks like the difference is when doing:
a) find_elements -> retrieve one -> action (ERROR)
b) find_element -> action (FINE)

**Comment by jsa34** [comment]
I just tried to compare the two elements in debugging through PyCharm, and it seemed to cause the JVM to crash.

I have attached the error file produced:

[hs_err_pid10932.log](https://github.com/gaozhao1989/pyjab/files/6771480/hs_err_pid10932.log)

**Comment by gaozhao1989** [comment]
Hi @jsa34 

Thank you for your kindly provide information.
Found root cause for this bug.
Element released before use JAB func, so it always returns -1.

jabelement.py => _get_elements_by_node
`
    if self._is_match_attributes(node_attributes, _jabelement):
       jabelements.append(_jabelement)
`
Need add continue to handle this condition

Another suggestion for find elements, you can directly use xpath like
`"//push button[@name='abc' and @indexinparent=2]"`

**Comment by jsa34** [comment]
Hello!

Thanks for getting back to me.

I have tried this, but getting the same result (adding a continue after:
 if self._is_match_attributes(node_attributes, _jabelement): jabelements.append(_jabelement)

I noticed another occurrence of this logic at _get_element_by_node that did a "break" instead of continue, so tried that but no luck :(

**Comment by gaozhao1989** [comment]
please check full code as below:
```
def _get_elements_by_node(
        self, node: str, level: str = "root", jabelement: JABElement = None
    ) -> list[JABElement]:
        """Get list of child JABElement by specific node

        Args:
            node (str): Node content for every single content in xpath.\n
            level (str, optional): Level for node, two options: "root" and "child". Defaults to "root".\n
            jabelement (JABElement, optional): The parent JABElement. Defaults to None.

        Raises:
            ValueError: Incorect level set

        Returns:
            list[JABElement]: list of the JABElement
        """
        dict_gen = {
            "root": self._generate_all_childs,
            "child": self._generate_childs_from_element,
        }
        if level not in dict_gen.keys():
            raise ValueError("level should be in 'root' or 'child'")
        node_info = self.xpath_parser.get_node_information(node)
        node_role = node_info.get("role")
        node_attributes = node_info.get("attributes")
        jabelement = self._get_node_element(jabelement)
        jabelements = list()
        for _jabelement in dict_gen[level](jabelement):
            if node_role not in ["*", _jabelement.role_en_us]:
                self.release_jabelement(_jabelement)
                continue
            if self._is_match_attributes(node_attributes, _jabelement):
                jabelements.append(_jabelement)
                continue  # <= set continue in here
            self.release_jabelement(_jabelement)
        return jabelements
```

if changes still not working, please try JAB internal func within loop 
```
 for _jabelement in dict_gen[level](jabelement):
    # try JAB internal func with "getObjectDepth" in here if can get correct callback
```

Otherwise you need try JAB internal func with click action
```
def click(self) -> None:
    """Clicks the JABElement."""
    self.win32_utils._set_window_foreground(hwnd=self.hwnd.value)

    acc_acts = AccessibleActions()
    self.bridge.getAccessibleActions(self.vmid, self.accessible_context, byref(acc_acts))

    act_todo = AccessibleActionsToDo()
    act_todo.actionsCount = acc_acts.actionsCount
    for i in range(0, act_todo.actionsCount):
        if  acc_acts.actionInfo[i].name.lower() == "click":
            act_todo.actions[i].name = acc_acts.actionInfo[i].name
            break

    self.bridge.doAccessibleActions(self.vmid, self.accessible_context, byref(act_todo), jint())
```

**Comment by gaozhao1989** [comment]
close with bug fixed in v1.0.1
